File: service-beacons-go/services/version1/BeaconsGrpcServiceV1.go
package services1

import (
	"context"

	logic "github.com/nov-pocs/samples/service-beacons-go/logic"
	protos "github.com/nov-pocs/samples/service-beacons-go/protos"
	cconv "github.com/pip-services3-go/pip-services3-commons-go/convert"
	cdata "github.com/pip-services3-go/pip-services3-commons-go/data"
	cref "github.com/pip-services3-go/pip-services3-commons-go/refer"
	cvalid "github.com/pip-services3-go/pip-services3-commons-go/validate"
	grpcservices "github.com/pip-services3-go/pip-services3-grpc-go/services"
)

type BeaconsGrpcServiceV1 struct {
	*grpcservices.GrpcService
	controller logic.IBeaconsController
}

func NewBeaconsGrpcServiceV1() *BeaconsGrpcServiceV1 {
	c := &BeaconsGrpcServiceV1{
		GrpcService: grpcservices.NewGrpcService("beacons_v1.BeaconsV1"),
	}
	c.GrpcService.IRegisterable = c
	c.DependencyResolver.Put("controller", cref.NewDescriptor("beacons", "controller", "default", "*", "*"))
	return c
}

func (c *BeaconsGrpcServiceV1) SetReferences(references cref.IReferences) {
	c.GrpcService.SetReferences(references)
	resolv, err := c.DependencyResolver.GetOneRequired("controller")
	if err == nil && resolv != nil {
		c.controller = resolv.(logic.IBeaconsController)
		return
	}
	panic("Can't resolve 'controller' reference")
}

func (c *BeaconsGrpcServiceV1) Open(correlationId string) error {
	// Add interceptors here
	return c.GrpcService.Open(correlationId)
}

func (c *BeaconsGrpcServiceV1) GetBeacons(
	ctx context.Context, req *protos.BeaconV1PageRequest) (*protos.BeaconV1PageReply, error) {

	filter := cdata.NewFilterParamsFromValue(req.GetFilter())
	paging := cdata.NewEmptyPagingParams()
	if req.Paging != nil {
		paging = cdata.NewPagingParams(req.Paging.GetSkip(), req.Paging.GetTake(), req.Paging.GetTotal())
	}
	page, err := c.controller.GetBeacons(
		req.CorrelationId,
		filter,
		paging,
	)

	result := protos.BeaconV1PageReply{
		Page: &protos.BeaconV1Page{},
	}

	if page.Total != nil {
		result.Page.Total = *page.Total
	}
	for _, v := range page.Data {
		buf := FromBeacon(v)
		result.Page.Data = append(result.Page.Data, buf)
	}
	return &result, err
}

func (c *BeaconsGrpcServiceV1) GetBeaconById(
	ctx context.Context, req *protos.BeaconV1IdRequest) (*protos.BeaconV1ObjectReply, error) {

	Schema := cvalid.NewObjectSchema().
		WithRequiredProperty("BeaconId", cconv.String)

	vErr := Schema.ValidateAndReturnError("", *req, false)
	result := protos.BeaconV1ObjectReply{}

	if vErr != nil {
		result.Error = FromError(vErr)
		return &result, vErr
	}

	beacon, err := c.controller.GetBeaconById(
		req.CorrelationId,
		req.BeaconId,
	)

	result.Error = FromError(err)
	result.Beacon = FromBeacon(beacon)
	return &result, err
}

func (c *BeaconsGrpcServiceV1) CreateBeacon(
	ctx context.Context, req *protos.BeaconV1ObjectRequest) (*protos.BeaconV1ObjectReply, error) {

	beacon := ToBeacon(req.Beacon)

	data, err := c.controller.CreateBeacon(
		req.CorrelationId,
		beacon,
	)

	result := &protos.BeaconV1ObjectReply{
		Beacon: FromBeacon(data),
		Error:  FromError(err),
	}

	return result, err
}

func (c *BeaconsGrpcServiceV1) UpdateBeacon(
	ctx context.Context, req *protos.BeaconV1ObjectRequest) (*protos.BeaconV1ObjectReply, error) {

	beacon := ToBeacon(req.Beacon)
	data, err := c.controller.UpdateBeacon(
		req.CorrelationId,
		beacon,
	)

	result := protos.BeaconV1ObjectReply{}
	result.Error = FromError(err)
	result.Beacon = FromBeacon(data)

	return &result, err
}

func (c *BeaconsGrpcServiceV1) DeleteBeaconById(
	ctx context.Context, req *protos.BeaconV1IdRequest) (*protos.BeaconV1ObjectReply, error) {

	Schema := cvalid.NewObjectSchema().
		WithRequiredProperty("BeaconId", cconv.String)

	vErr := Schema.ValidateAndReturnError("", *req, false)
	result := protos.BeaconV1ObjectReply{}

	if vErr != nil {
		result.Error = FromError(vErr)
		return &result, vErr
	}

	beacon, err := c.controller.DeleteBeaconById(
		req.CorrelationId,
		req.BeaconId,
	)

	result.Error = FromError(err)
	result.Beacon = FromBeacon(beacon)
	return &result, err
}

func (c *BeaconsGrpcServiceV1) GetBeaconByUdi(
	ctx context.Context, req *protos.BeaconV1UdiRequest) (*protos.BeaconV1ObjectReply, error) {

	Schema := cvalid.NewObjectSchema().
		WithRequiredProperty("BeaconUdi", cconv.String)

	vErr := Schema.ValidateAndReturnError("", *req, false)
	result := protos.BeaconV1ObjectReply{}

	if vErr != nil {
		result.Error = FromError(vErr)
		return &result, vErr
	}

	beacon, err := c.controller.GetBeaconByUdi(
		req.CorrelationId,
		req.BeaconUdi,
	)

	result.Error = FromError(err)
	result.Beacon = FromBeacon(beacon)
	return &result, err
}

func (c *BeaconsGrpcServiceV1) CalculatePosition(
	ctx context.Context, req *protos.BeaconV1PositionRequest) (*protos.BeaconV1PositionReply, error) {

	Schema := cvalid.NewObjectSchema().
		WithRequiredProperty("SiteId", cconv.String).
		WithRequiredProperty("Udis", cvalid.NewArraySchema(cconv.String))

	vErr := Schema.ValidateAndReturnError("", *req, false)
	result := protos.BeaconV1PositionReply{}

	if vErr != nil {
		result.Error = FromError(vErr)
		return &result, vErr
	}

	position, err := c.controller.CalculatePosition(
		req.CorrelationId,
		req.SiteId,
		req.Udis,
	)

	result.Error = FromError(err)
	result.Point = FromPosition(position)
	return &result, err
}

func (c *BeaconsGrpcServiceV1) Register() {
	protos.RegisterBeaconsV1Server(c.Endpoint.GetServer(), c)
}

File: service-beacons-dotnet/src/Service/Services/Version1/BeaconsGrpcServiceV1.cs
using Nov.MaxSamples.Beacons.Logic;
using BeaconsV1;
using Grpc.Core;
using PipServices3.Commons.Data;
using PipServices3.Commons.Data.Mapper;
using PipServices3.Commons.Errors;
using PipServices3.Commons.Refer;
using PipServices3.Grpc.Services;
using System;
using System.Linq;
using System.Threading.Tasks;

namespace Nov.MaxSamples.Beacons.Services.Version1
{
	public class BeaconsGrpcServiceV1 : GrpcService
	{
		private IBeaconsController _controller;

		public BeaconsGrpcServiceV1()
			: this("beacons_v1")
		{
		}

		public BeaconsGrpcServiceV1(string name)
			: base(name)
		{
			_dependencyResolver.Put("controller", new Descriptor("beacons", "controller", "default", "*", "*"));
		}

		public override void SetReferences(IReferences references)
		{
			base.SetReferences(references);
			_controller = _dependencyResolver.GetOneRequired<IBeaconsController>("controller");
		}

		private async Task<BeaconV1PageReply> GetBeaconsAsync(BeaconV1PageRequest request, ServerCallContext context)
		{
			var correlationId = request.CorrelationId;
			var methodName = _serviceName + ".get_beacons";

			var response = new BeaconV1PageReply();

			var filter = new FilterParams(request.Filter);
			var paging = new PipServices3.Commons.Data.PagingParams(request.Paging.Skip, request.Paging.Take, request.Paging.Total);

			try
			{
				using (var timing = Instrument(correlationId, methodName))
				{
					var result = await _controller.GetBeaconsAsync(
						correlationId,
						filter,
						paging
					);

					response.Page = BeaconsGrpcConverterV1.FromBeaconsPage(result);
				}
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				response.Error = ObjectMapper.MapTo<BeaconsV1.ErrorDescription>(ErrorDescriptionFactory.Create(ex));
			}

			return response;
		}

		private async Task<BeaconV1ObjectReply> GetBeaconByIdAsync(BeaconV1IdRequest request, ServerCallContext context)
		{
			var correlationId = request.CorrelationId;
			var methodName = _serviceName + ".get_beacon_by_id";

			var response = new BeaconV1ObjectReply();

			try
			{
				using (var timing = Instrument(correlationId, methodName))
				{
					var result = await _controller.GetBeaconByIdAsync(
						correlationId,
						request.BeaconId
					);

					response.Beacon = BeaconsGrpcConverterV1.FromBeacon(result);
				}
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				response.Error = ObjectMapper.MapTo<BeaconsV1.ErrorDescription>(ErrorDescriptionFactory.Create(ex));
			}

			return response;
		}

		private async Task<BeaconV1ObjectReply> GetBeaconByUdiAsync(BeaconV1UdiRequest request, ServerCallContext context)
		{
			var correlationId = request.CorrelationId;
			var methodName = _serviceName + ".get_beacon_by_id";

			var response = new BeaconV1ObjectReply();

			try
			{
				using (var timing = Instrument(correlationId, methodName))
				{
					var result = await _controller.GetBeaconByUdiAsync(
						correlationId,
						request.BeaconUdi
					);

					response.Beacon = BeaconsGrpcConverterV1.FromBeacon(result);
				}
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				response.Error = ObjectMapper.MapTo<BeaconsV1.ErrorDescription>(ErrorDescriptionFactory.Create(ex));
			}

			return response;
		}

		private async Task<BeaconV1PositionReply> CalculatePositionAsync(BeaconV1PositionRequest request, ServerCallContext context)
		{
			var correlationId = request.CorrelationId;
			var methodName = _serviceName + ".calculate_position";

			var response = new BeaconV1PositionReply();

			try
			{
				using (var timing = Instrument(correlationId, methodName))
				{
					var result = await _controller.CalculatePositionAsync(
						correlationId,
						request.SiteId,
						request.Udis.ToArray()
					);

					response.Point = BeaconsGrpcConverterV1.FromPosition(result);
				}
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				response.Error = ObjectMapper.MapTo<BeaconsV1.ErrorDescription>(ErrorDescriptionFactory.Create(ex));
			}

			return response;
		}

		private async Task<BeaconV1ObjectReply> CreateBeaconAsync(BeaconV1ObjectRequest request, ServerCallContext context)
		{
			var correlationId = request.CorrelationId;
			var methodName = _serviceName + ".create_beacon";

			var response = new BeaconV1ObjectReply();

			try
			{
				using (var timing = Instrument(correlationId, methodName))
				{
					var result = await _controller.CreateBeaconAsync(
						correlationId,
						BeaconsGrpcConverterV1.ToBeacon(request.Beacon)
					);

					response.Beacon = BeaconsGrpcConverterV1.FromBeacon(result);
				}
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				response.Error = ObjectMapper.MapTo<BeaconsV1.ErrorDescription>(ErrorDescriptionFactory.Create(ex));
			}

			return response;
		}

		private async Task<BeaconV1ObjectReply> UpdateBeaconAsync(BeaconV1ObjectRequest request, ServerCallContext context)
		{
			var correlationId = request.CorrelationId;
			var methodName = _serviceName + ".update_beacon";

			var response = new BeaconV1ObjectReply();

			try
			{
				using (var timing = Instrument(correlationId, methodName))
				{
					var result = await _controller.UpdateBeaconAsync(
						correlationId,
						BeaconsGrpcConverterV1.ToBeacon(request.Beacon)
					);

					response.Beacon = BeaconsGrpcConverterV1.FromBeacon(result);
				}
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				response.Error = ObjectMapper.MapTo<BeaconsV1.ErrorDescription>(ErrorDescriptionFactory.Create(ex));
			}

			return response;
		}


		private async Task<BeaconV1ObjectReply> DeleteBeaconByIdAsync(BeaconV1IdRequest request, ServerCallContext context)
		{
			var correlationId = request.CorrelationId;
			var methodName = _serviceName + ".delete_beacon_by_id";

			var response = new BeaconV1ObjectReply();

			try
			{
				using (var timing = Instrument(correlationId, methodName))
				{
					var result = await _controller.DeleteBeaconByIdAsync(
						correlationId,
						request.BeaconId
					);

					response.Beacon = BeaconsGrpcConverterV1.FromBeacon(result);
				}
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				response.Error = ObjectMapper.MapTo<BeaconsV1.ErrorDescription>(ErrorDescriptionFactory.Create(ex));
			}

			return response;
		}

		protected override void OnRegister()
		{
			RegisterMethod<BeaconV1PageRequest, BeaconV1PageReply>("get_beacons", GetBeaconsAsync);
			RegisterMethod<BeaconV1IdRequest, BeaconV1ObjectReply>("get_beacon_by_id", GetBeaconByIdAsync);
			RegisterMethod<BeaconV1UdiRequest, BeaconV1ObjectReply>("get_beacon_by_udi", GetBeaconByUdiAsync);
			RegisterMethod<BeaconV1PositionRequest, BeaconV1PositionReply>("calculate_position", CalculatePositionAsync);
			RegisterMethod<BeaconV1ObjectRequest, BeaconV1ObjectReply>("create_beacon", CreateBeaconAsync);
			RegisterMethod<BeaconV1ObjectRequest, BeaconV1ObjectReply>("update_beacon", UpdateBeaconAsync);
			RegisterMethod<BeaconV1IdRequest, BeaconV1ObjectReply>("delete_beacon_by_id", DeleteBeaconByIdAsync);
		}
	}
}
